sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala

```diff
@@ -115,18 +115,32 @@ case class OptimizeSkewedJoin(conf: SQLConf) extends Rule[SparkPlan] {
     partitionStartIndices += 0
     var i = 0
     var postMapPartitionSize = 0L
+    var lastPackagedPartitionSize = -1L
     while (i < mapPartitionSizes.length) {
       val nextMapPartitionSize = mapPartitionSizes(i)
       if (i > 0 && postMapPartitionSize + nextMapPartitionSize > targetSize) {
         partitionStartIndices += i
+        lastPackagedPartitionSize = postMapPartitionSize
         postMapPartitionSize = nextMapPartitionSize
       } else {
         postMapPartitionSize += nextMapPartitionSize
       }
       i += 1
     }
 
-    partitionStartIndices.toArray
+    if (lastPackagedPartitionSize > -1) {
+      val lastPartitionDiff = math.abs(targetSize - postMapPartitionSize)
+      val diffIfMergeLastPartition = math.abs(
+        lastPackagedPartitionSize + postMapPartitionSize - targetSize)
+      // If the last partition is very small, we should merge it to the previous partition.
+      if (lastPartitionDiff > diffIfMergeLastPartition * 2) {
```

Comment:
I get your point here, but using target size and this formula doesn't quite make sense sometimes, e.g.,
```
targetSize = 7
lastButOneSize = 4
lastSize = 4
```
You'd get:
```
diffLastPartition = 3
diffLastTwoPartitions = 1
```
This would satisfy your "very small" condition, yet I'd argue that first of all the last partition is well over half the target size, so I wouldn't consider it too small; second, what if previous partitions are mostly around size of 4 too? It would more even not to merge them, right?

A simple way is probably get the average size of all partitions except the last one and merge if `(avgSize - lastSize) > (lastSize + lastButOneSize - avgSize)`.

Reply:
Here, we may also need consider whether the` (lastSize + lastButOnesize)` is larger than the `targetSize`. 
`((avgSize - lastSize) > (lastSize + lastButOneSize - avgSize)) && (lastSize + lastButOneSize) < targetSize`

Reply:
Shouldn't it always be larger than targetSize?? otherwise the two splits would have been one in the first place.

Reply:
Good discussion here! I'm thinking about more cases:
```
targetSize = 7
lastButOneSize = 1
lastSize = 7
```
shall we merge the last partition into the previous partition?

Maybe we can use a simple heuristic: merge 2 adjacent partitions if they don't exceed the target size too much (say 20%?). And we can apply it to partitions in the middle, not have to be the last 2 partitions.

Reply:
Yes, this would work for extreme cases where you could have a very small split in between two large splits very close to the target size.